cocos/navmesh/CCNavMesh.cpp

```diff
@@ -0,0 +1,637 @@
+ copies of the Software, and to permit persons to whom the Software is
+ furnished to do so, subject to the following conditions:
+ 
+ The above copyright notice and this permission notice shall be included in
+ all copies or substantial portions of the Software.
+ 
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+ ****************************************************************************/
+#include "navmesh/CCNavMesh.h"
+#if CC_USE_NAVMESH
+
+#include "platform/CCFileUtils.h"
+#include "renderer/CCRenderer.h"
+#include "recast/Detour/DetourCommon.h"
+#include "recast/DebugUtils/DetourDebugDraw.h"
+#include <sstream>
+
+NS_CC_BEGIN
+
+struct TileCacheSetHeader
+{
+    int magic;
+    int version;
+    int numTiles;
+    dtNavMeshParams meshParams;
+    dtTileCacheParams cacheParams;
+};
+
+struct TileCacheTileHeader
+{
+    dtCompressedTileRef tileRef;
+    int dataSize;
+};
+
+static unsigned char* parseRow(unsigned char* buf, unsigned char* bufEnd, char* row, int len)
+{
+    bool start = true;
+    bool done = false;
+    int n = 0;
+    while (!done && buf < bufEnd)
+    {
+        char c = *buf;
+        buf++;
+        // multirow
+        switch (c)
+        {
+        case '\n':
+            if (start) break;
+            done = true;
+            break;
+        case '\r':
+            break;
+        case '\t':
+        case ' ':
+            if (start) break;
+        default:
+            start = false;
+            row[n++] = c;
+            if (n >= len - 1)
+                done = true;
+            break;
+        }
+    }
+    row[n] = '\0';
+    return buf;
+}
+
+static const int TILECACHESET_MAGIC = 'T' << 24 | 'S' << 16 | 'E' << 8 | 'T'; //'TSET';
+static const int TILECACHESET_VERSION = 1;
+static const int MAX_AGENTS = 128;
+
+NavMesh* NavMesh::create(const std::string &navFilePath, const std::string &geomFilePath)
+{
+    auto ref = new (std::nothrow) NavMesh();
+    if (ref && ref->initWithFilePath(navFilePath, geomFilePath))
+    {
+        ref->autorelease();
+        return ref;
+    }
+    CC_SAFE_DELETE(ref);
+    return nullptr;
+}
+
+NavMesh::NavMesh()
+    : _navMesh(nullptr)
+    , _navMeshQuery(nullptr)
+    , _crowed(nullptr)
+    , _tileCache(nullptr)
+    , _allocator(nullptr)
+    , _compressor(nullptr)
+    , _meshProcess(nullptr)
+    , _geomData(nullptr)
+    , _isDebugDrawEnabled(false)
+{
+
+}
+
+NavMesh::~NavMesh()
+{
+    dtFreeTileCache(_tileCache);
+    dtFreeCrowd(_crowed);
+    dtFreeNavMesh(_navMesh);
+    dtFreeNavMeshQuery(_navMeshQuery);
```

Comment:
@ricardoquesada  

We are using third library for path finding. The recast, which also contain a tool to build the navigation mesh offline. 

`dt` is the prefix of recast.